transcom/mymove
# [MB-13908] Update golangci-lint config

## [Jira ticket](https://dp3.atlassian.net/browse/MB-13908) for this change

## Summary

This PR updates and cleans up our `golangci-lint` config.  Some of this work was deferred from #9064 when we added the `gci` linter.  Specifically, the following changes are being made:

- three linters that are now deprecated are being removed: `deadcode`, `structcheck`, and `varcheck`.  [Per the documentation](https://golangci-lint.run/usage/linters/#disabled-by-default), these linters were abandoned by the author and the `unused` linter should be used as the replacement for all three.
- the list of linters noted in the `enable` section was not the complete list that were being run.  There are several others that were [enabled by default](https://golangci-lint.run/usage/linters/#enabled-by-default), including the `unused` one noted above.  So, I've added `gosimple` and `unused` in explicitly here to avoid any confusion about what linters are actually running.  If you want to see a list of all enabled linters, get a local version of `golangci-lint` and run `golangci-lint linters`.
- the `deadline` setting under `run` has been deprecated and [`timeout` should be used in its place](https://github.com/golangci/golangci-lint/pull/793).
- instead of just skipping `mocks` directories, I skip any directory that ends in `mocks` since we do have a mocks directory called `ghcmocks` that we had to create to get around some circular dependency issues.
- the workaround of setting a `go` version to `1.17` that we introduced when [moving to go 1.18](https://github.com/transcom/mymove/pull/8327) has been removed because the problem linters have either been fixed or deprecated.
- in removing the `1.17` workaround, I realized that [an annotation and linter suppression added/approved in a previous PR](https://github.com/transcom/mymove/pull/8327/files#r835386427) was inadvertently removed in a subsequent PR's refactor.  This PR adds it back.
- the `CODEOWNERS` file was referencing the wrong filename for the golangci-lint config, so I've fixed that.

## Setup to Run Your Code

How to test:
- `make clean server_build client_build build_tools`
- `pre-commit run -a` to make sure there are no linter issues
- `pre-commit run -a --verbose golangci-lint` to see if there are any warnings about linters being turned off (there shouldn't be anymore)
- `golangci-lint linters` to see a list of enabled/disabled linters for our configuration

## Verification Steps for Author

These are to be checked by the author.
